File: mirea/cources/2026-2027/aii-august/04/contest/C.cpp
#include <bits/stdc++.h>
#include <cassert>
using namespace std;
const int mod = (int)1e9+7;
// Быстрое возведение в степень:
int binpow(int a, int n) {
    int r = 1;
    while (n > 0) {
        if (n % 2 == 1)
            r = int(r * 1LL * a % mod);
        a = int(a * 1LL * a % mod);
        n >>= 1;
    }
    return r;
}
// Факториалы и обратные к ним:
const int NMAX = 1 << 21;
int fact[NMAX], ifact[NMAX];
void precalc() {
    fact[0] = 1;
    for (int i = 1; i < NMAX; i++)
        fact[i] = int(fact[i-1] * 1LL * i % mod);
    ifact[NMAX-1] = binpow(fact[NMAX-1], mod-2);
    for (int i = NMAX-2; i >= 0; i--)
        ifact[i] = int((i+1LL) * ifact[i+1] % mod);
}
// Биномиальный коэффициент:
int C(int n, int k) {
    if (n < 0 || k < 0 || k > n)
        return 0;
    return int(fact[n] * 1LL * ifact[k] % mod * ifact[n-k] % mod);
}
// Решение:
void solve() {
    int n, d, q; cin >> n >> d >> q;
    map<int, int> cnt;
    for (int i = 0; i < n; i++) {
        int a, c; cin >> a >> c;
        cnt[a] += c;
    }
    while (q --> 0) {
        char ch; cin >> ch;
        if (ch == '+') {
            int a, c; cin >> a >> c;
            cnt[a] += c;
        } else if (ch == '-') {
            int a, c; cin >> a >> c;
            cnt[a] -= c;
            if (cnt[a] == 0)
                cnt.erase(cnt.find(a));
        } else {
            assert(ch == '?');
            int k; cin >> k;
            cout << C(cnt.size() + k - 1, k) << '\n';
        }
    }
}
main() {
    precalc(); // вызов предподсчёта факториалов и обратных к ним
    int tt; cin >> tt;
    while (tt --> 0) solve();
}

File: mirea/cources/2026-2027/aii-august/04/contest/C.py
from collections import defaultdict
mod = int(1e9) + 7
# Факториалы и обратные к ним:
NMAX = 1 << 21
fact, ifact = [0] * NMAX, [0] * NMAX
def precalc():
    fact[0] = 1
    for i in range(1, NMAX):
        fact[i] = fact[i - 1] * i % mod
    ifact[NMAX - 1] = pow(fact[NMAX - 1], mod - 2, mod)
    for i in range(NMAX - 2, -1, -1):
        ifact[i] = (i + 1) * ifact[i + 1] % mod
# Биномиальный коэффициент:
def C(n, k):
    if n < 0 or k < 0 or k > n:
        return 0
    return fact[n] * ifact[k] % mod * ifact[n - k] % mod
# Решение:
def solve():
    n, d, q = map(int, input().split())
    cnt = defaultdict(int)
    for i in range(n):
        a, c = map(int, input().split())
        cnt[a] += c
    while q > 0:
        ch = input().split()
        if ch[0] == '+':
            a, c = map(int, ch[1:])
            cnt[a] += c
        elif ch[0] == '-':
            a, c = map(int, ch[1:])
            cnt[a] -= c
            if cnt[a] == 0:
                del cnt[a]
        else:
            assert ch[0] == '?'
            k = int(ch[1])
            print(C(len(cnt) + k - 1, k))
        q -= 1
precalc() # вызов предподсчёта факториалов и обратных к ним
tt = int(input())
while tt > 0:
    solve()
    tt -= 1